test/datasets_utils.py

```diff
@@ -0,0 +1,614 @@
+                    del special_kwargs["download"]
+
+    def test_not_found(self):
+        with self.assertRaises(RuntimeError):
+            with self.create_dataset(inject_fake_data=False):
+                pass
+
+    def test_smoke(self):
+        with self.create_dataset() as (dataset, _):
+            self.assertIsInstance(dataset, torchvision.datasets.VisionDataset)
+
+    @test_all_configs
+    def test_str_smoke(self, config):
+        with self.create_dataset(config) as (dataset, _):
+            self.assertIsInstance(str(dataset), str)
+
+    @test_all_configs
+    def test_feature_types(self, config):
+        with self.create_dataset(config) as (dataset, _):
+            example = dataset[0]
+
+            actual = len(example)
+            expected = len(self.FEATURE_TYPES)
+            self.assertEqual(
+                actual,
+                expected,
+                f"The number of the returned features does not match the the number of elements in in FEATURE_TYPES: "
+                f"{actual} != {expected}",
+            )
+
+            for idx, (feature, expected_feature_type) in enumerate(zip(example, self.FEATURE_TYPES)):
+                with self.subTest(idx=idx):
+                    self.assertIsInstance(feature, expected_feature_type)
+
+    @test_all_configs
+    def test_num_examples(self, config):
+        with self.create_dataset(config) as (dataset, info):
+            self.assertEqual(len(dataset), info["num_examples"])
+
+
+class ImageDatasetTestCase(DatasetTestCase):
+    """Abstract base class for image dataset testcases.
+
+    - Overwrites the FEATURE_TYPES class attribute to expect a :class:`PIL.Image.Image` and an integer label.
+    """
+
+    FEATURE_TYPES = (PIL.Image.Image, int)
+
+    @contextlib.contextmanager
+    def create_dataset(
+        self,
+        config: Optional[Dict[str, Any]] = None,
+        inject_fake_data: bool = True,
+        disable_download_extract: Optional[bool] = None,
+        **kwargs: Any,
+    ) -> Iterator[Tuple[torchvision.datasets.VisionDataset, Dict[str, Any]]]:
+        with super().create_dataset(
+            config=config,
+            inject_fake_data=inject_fake_data,
+            disable_download_extract=disable_download_extract,
+            **kwargs,
+        ) as (dataset, info):
+            # PIL.Image.open() only loads the image meta data upfront and keeps the file open until the first access
+            # to the pixel data occurs. Trying to delete such a file results in an PermissionError on Windows. Thus, we
+            # track all lazily opened images and close the file handle before the file is deleted.
+            # This problem only occurs during testing since some tests, e.g. DatasetTestCase.test_feature_types open an
+            # image, but never use the underlying data. During normal operation it is reasonable to assume that the
+            # user wants to work with the image he just opened rather than deleting the underlying file.
```

Comment:
AFAIK `PIL.Image.open().load()` would be enough. No need to force a specific image mode such as `RGB`. My rationale was that this is a change a contributor needs to implement in the dataset and we need to spot this during code review. Given that we want to make adding new datasets easier for the contributor and us, I'm not sure if we want to switch to this "manually". Especially, since this has no relevance in real applications as explained in the comment.